### PR TITLE
feat: precompile crypto on a syscall-backed Crypto provider (FLU-1456)

### PR DESCRIPTION
**Pull request by dmitry123**
Parent task: FLU-1456. One commit per sub-task, plus the review and decision trail that the branch history keeps (no force-pushes):

1. **test(runtime,e2e)** 142cb938e — FLU-1457. Reference tests for every tower and weierstrass syscall handler against arkworks, k256 and p256, pinning the SP1 rules (equal points and non-canonical coordinates rejected, raw-integer subtraction underflow, zero slope denominator inverting to zero). A differential e2e suite runs every precompile vector natively through `revm::precompile` and as a transaction against the guest and requires identical output and total gas, or failure on both sides: the EIP-2537 suites, bn256, ecrecover, p256verify, KZG, modexp, blake2f, sha256, ripemd160, identity. The measurement tests (BLS throughput, per-precompile survey, reload cost, syscall round trips; extended in f9b766111) are ignored by default and have a manual `precompile-survey` workflow.
2. **perf(runtime)** 06de7fc75 — FLU-1458. The tower and weierstrass handlers move from `num::BigUint` / `sp1_curves::AffinePoint` to fixed-width arkworks fields behind the unchanged ABI. `_tower_fp1_bls12381_mul` round trip 464 ns → 111 ns; the bn256 `ecAdd` guest, the only genesis user of these syscalls, 3.5 → 47 Mgas/s.
3. **feat(runtime,types)** b6d2ee2af — FLU-1459. A precompile-level syscall family shaped like revm-precompile's `Crypto` trait (`_crypto_bls12381_pairing_check`, `_crypto_bls12381_g1_msm`, `_crypto_secp256k1_ecrecover`, `_crypto_kzg_verify_proof`, …), uniform `(input_ptr, input_len, output_ptr) -> code` ABI with the trait's own byte encoding, host handlers over `DefaultCrypto` (blst, k256, c-kzg). Engine-metered callers pay the EIP gas of the payload in the handler, before the input is copied; self-metered system runtimes are not charged because the guest charges the same gas through `sync_evm_gas` (`RuntimeContext::engine_metered`). Indices are appended, nothing existing changes. First landed as d6232cc2e, reverted in 6c746ceac while the alternative below was measured, reinstated with the review fixes from 2e576b25d and the crates.io version pin on `revm-precompile` that `cargo package` needs for the published runtime crate.
4. **feat(contracts,crypto)** b65ffd81e — FLU-1442. `fluentbase_crypto::PrecompileCrypto`, behind the `revm-precompile` feature of `fluentbase-crypto`, implements the trait over the syscalls; the bls12381, bn256, ecrecover, eip7951 and kzg guests install it at entry and keep calling the unchanged `revm_precompile` functions. First landed as 5f167cee3 with a separate crate, reverted in 340e5a576, reinstated without it.
5. **build(e2e)** b9089eee0 — lockfile-only: the standalone `e2e/evm` and `e2e/codec` suites carry their own `Cargo.lock`, refreshed so their `--locked` CI jobs resolve.

### Why a new syscall family and not the existing SP1-shaped syscalls

The existing tower and curve syscalls stay untouched, and a guest built on them was measured before the family was reinstated (rwasm-patches `bls12_381`, pairing only). Per pairing pair on the same machine: 1.37 ms with arkworks inside the guest, 3.17 ms on the fine-grained syscalls (1.76 ms with the subgroup checks removed), 0.20 ms on `_crypto_bls12381_pairing_check`; in the interpreter 508 ms, 97 ms and 0.18 ms. The fine-grained route is bounded by the host round trip per field operation (fp1 mul 111 ns, fp2 mul 236 ns, of which the bare call is 8 ns and the rest operand copies and Montgomery conversions) and by a host inversion per affine point operation (`_bls12381_double` 6.3 µs), which the subgroup checks hit 128 times per point. That granularity is right inside a zkVM, where a syscall is a chip row, and wrong on a node. A zkVM host implements the same trait behind the same ABI (openvm-eth already does; SP1 through its patched crates).

### Numbers (this machine, release)

| call | before, wasmtime | after, wasmtime | before, interpreter | after, interpreter |
|---|---|---|---|---|
| BLS pairing, 512 pairs (16.7M gas) | 0.70 s | 0.09 s | 260 s | 0.09 s |
| BLS pairing, 3066 pairs (100M gas) | 4.2 s | 0.58 s | — | — |
| BLS G2 MSM, 1423 points | 0.33 s | 0.05 s | ~99 s | 0.05 s |
| ecrecover | 13 Mgas/s | 144 Mgas/s | | 71 Mgas/s |
| KZG point evaluation | 14 Mgas/s | 73 Mgas/s | | 74 Mgas/s |
| p256verify | 14 Mgas/s | 44 Mgas/s | | 41 Mgas/s |
| bn256 pairing, 2 pairs | 74 Mgas/s | 224 Mgas/s | | 229 Mgas/s |
| bn256 add | 3.5 Mgas/s | 47 Mgas/s | | 5 Mgas/s |

### Verification (final state)

- `cargo test -p fluentbase-types --lib`: 22; `cargo test -p fluentbase-runtime --features std --lib`: 117.
- `cargo test -p fluentbase-e2e --release` for `crypto_syscalls`, `precompile_vectors`, the `bls12381` tests and the survey: 33 green under wasmtime, 29 under the interpreter, identical gas to the native precompiles.
- Guests' native unit tests in the contracts workspace: 55 green on the syscall provider.
- clippy `-D warnings` on the touched root crates and the contracts workspace; rustfmt clean; `cargo package -p fluentbase-crypto` and `-p fluentbase-runtime` pass; `cargo fetch --locked` for contracts, e2e/evm, e2e/codec and examples.

### Not in this PR

- Runtime-upgrade payloads for devnet, testnet and mainnet, and the historical-payload admission replay before deploy.
- Removing the now-dead in-wasm arkworks / k256 / p256 code from the guest builds (needs a revm-precompile feature to drop `DefaultCrypto`).
- Sub-tasks 5 (zkVM implementations) and 6 (retiring unused syscalls, `legacy.rs`).

## Summary by CodeRabbit

- **New Features**
  - Added crypto syscall support for BLS12-381, BN254, secp256k1, secp256r1, and KZG operations.
  - Added host-backed execution for supported cryptographic precompiles.
  - Added consistent input validation, output handling, error reporting, and EIP-aligned gas accounting.
  - Added support for invoking these operations from contracts and system runtimes.

- **Bug Fixes**
  - Improved elliptic-curve arithmetic handling, including malformed coordinates and edge cases.

- **Tests**
  - Added comprehensive compatibility, correctness, gas, throughput, and syscall-cost coverage.

### CODE REVIEW

File: .github/workflows/precompile-survey.yml [hunk 1]
```diff
@@ -0,0 +1,29 @@
...
+      - name: Install Rust toolchain
+        uses: dtolnay/rust-toolchain@fa04a1451ff1842e2626ccb99004d0195b455a88 # master
+        with:
+          toolchain: stable
+          targets: wasm32-unknown-unknown
+      - name: Run the survey (${{ matrix.features }})
+        run: |
+          cargo test -p fluentbase-e2e --release --no-default-features --features ${{ matrix.features }} \
+            precompile_throughput_survey bls_fp_mul_syscall_round_trip system_runtime_reload_cost \
+            -- --ignored --nocapture --test-threads=1
```
[the remark this answers is not in the document]
Comment (dmitry123) [reply .github/workflows/precompile-survey.yml]: Moved the three test filters after `--`, where libtest accepts several of them; still a single invocation. Fixed in 2e576b25d.

File: crates/precompile-crypto/src/lib.rs [hunk 2]
```diff
@@ -0,0 +1,211 @@
...
+
+/// Installs [`SyscallCrypto`] as the process-wide provider. Returns `false` if a provider was
+/// already installed, which is the normal case on every entry after the first.
+pub fn install() -> bool {
+    install_crypto(SyscallCrypto)
+}
+
+fn call(op: CryptoSyscall, input: &[u8]) -> Result<Vec<u8>, PrecompileHalt> {
+    let mut output = alloc::vec![0u8; op.output_len()];
+    match CryptoRuntime::crypto_syscall(op, input, &mut output) {
```
[the remark this answers is not in the document]
Comment (dmitry123) [reply crates/precompile-crypto/src/lib.rs]: Dropped the 4 MiB ceiling instead of propagating it: every payload byte is charged at the EIP rate before it reaches the host (engine-metered callers before the copy, precompile guests through their own gas accounting), so the length is bounded by the caller's gas and no aligned input is rejected. Fixed in 2e576b25d.

File: crates/runtime/src/syscall_handler/crypto.rs [hunk 3]
```diff
@@ -0,0 +1,479 @@
...
+    result: &mut [Value],
+) -> Result<(), TrapCode> {
+    let input_ptr = params[0].i32().unwrap() as u32 as usize;
+    let input_len = params[1].i32().unwrap() as u32 as usize;
+    let output_ptr = params[2].i32().unwrap() as u32 as usize;
+    if !op.accepts_input_len(input_len) {
+        result[0] = Value::I32(CryptoSyscallError::InvalidInput as i32);
+        return Ok(());
+    }
+    let input = caller.memory_read_into_vec(input_ptr, input_len)?;
```
[the remark this answers is not in the document]
Comment (dmitry123) [reply crates/runtime/src/syscall_handler/crypto.rs]: Fuel is now charged before the guest input is copied. Fixed in 2e576b25d.

File: crates/types/src/rwasm_context.rs [hunk 4]
```diff
@@ -131,6 +132,26 @@ impl NativeAPI for RwasmContext {
 
 #[rustfmt::skip]
 impl CryptoAPI for RwasmContext {
+    #[inline(always)]
+    fn crypto_syscall(op: CryptoSyscall, input: &[u8], output: &mut [u8]) -> i32 {
+        debug_assert_eq!(output.len(), op.output_len());
```
[the remark this answers is not in the document]
Comment (dmitry123) [reply crates/types/src/rwasm_context.rs]: Replaced the debug assertion with a runtime check that returns `InvalidInput`. Fixed in 2e576b25d.

File: e2e/runtime/src/bls12381.rs [hunk 5]
```diff
@@ -0,0 +1,369 @@
+//! Reproduces the multi-second BLS12-381 (EIP-2537) precompile calls.
+//!
+//! All seven BLS addresses run one guest rWasm module (`contracts/bls12381`) built from
+//! `revm-precompile` with default features off, so the curve arithmetic is pure-Rust arkworks
+//! executed by the system runtime: through Wasmtime with the `wasmtime` feature, through the
+//! rWasm interpreter without it. Neither path has host-side acceleration, so the cost is the
+//! execution itself and grows linearly with the number of pairs or points. Reloading the module
+//! after a runtime-cache reset is not the problem: the rWasm compile costs about 10 ms and the
+//! Wasmtime artifact is cached process-wide.
```
[the remark this answers is not in the document]
Comment (dmitry123) [reply e2e/runtime/src/bls12381.rs]: Rewrote the module docs for the syscall-backed path and refreshed the measurements. Fixed in 2e576b25d.

**Comment by dmitry123** [comment]
Testing wide WASM proposal, blocked for now
